### PR TITLE
[FLINK-21685][k8s] Introduce dedicated thread pool and cache for Kubernetes client

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

Currently, we use the same executor for `KubernetesRunningJobsRegistry.writeJobStatusToConfigMap` and for the `Fabric8FlinkKubeClient`, which is `io-executor`. This design may cause dead lock.

This PR tries to introduce a dedicated thread pool for Kubernetes client IO operations. Also this PR enable the CACHE for Kubernetes client, which means the `KubernetesResourceManager` and `KubernetesHAService` will share the same Kubernetes client.

## Brief change log

* Use dedicated thread pool for Kubernetes client IO operations
* Add cache for Flink kubernetes client

## Verifying this change

* Add new unit tests
* Verified in a minikube with application mode

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClientFactory.java [hunk 1]
```diff
@@ -87,12 +89,23 @@ public FlinkKubeClient fromConfiguration(Configuration flinkConfig, Executor ioE
...
+                        flinkConfig, client, () -> createThreadPoolForAsyncIO(poolSize));
+        CACHE.put(cacheKey, flinkKubeClient);
+        return flinkKubeClient;
     }
 
-    private static Executor createThreadPoolForAsyncIO() {
-        return Executors.newFixedThreadPool(2, new ExecutorThreadFactory("FlinkKubeClient-IO"));
+    private static ExecutorService createThreadPoolForAsyncIO(int poolSize) {
+        return Executors.newFixedThreadPool(
+                poolSize, new ExecutorThreadFactory("flink-kubeclient-io"));
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClientFactory.java:109]: If we create multiple `FlinkKubeClients`, then we might wanna give their thread pools different names.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for creating this PR @wangyang0918. The overall change looks good. What I am not so sure about is the caching mechanism because I think in its current form it might break.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClientFactory.java [hunk 2]
```diff
@@ -83,11 +89,19 @@ public FlinkKubeClient fromConfiguration(Configuration flinkConfig) {
...
         final NamespacedKubernetesClient client = new DefaultKubernetesClient(config);
         final int poolSize =
                 flinkConfig.get(KubernetesConfigOptions.KUBERNETES_CLIENT_IO_EXECUTOR_POOL_SIZE);
-        return new Fabric8FlinkKubeClient(
-                flinkConfig, client, () -> createThreadPoolForAsyncIO(poolSize));
+        final FlinkKubeClient flinkKubeClient =
+                new Fabric8FlinkKubeClient(
+                        flinkConfig, client, () -> createThreadPoolForAsyncIO(poolSize));
+        CACHE.put(cacheKey, flinkKubeClient);
+        return flinkKubeClient;
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/FlinkKubeClientFactory.java:104]: Thanks for your timely comments. Actually, I am still refining this PR about the second commit. The changes are not only about moving the `CACHE` from `FlinkKubeClientFactory`(of cause it is not a good place) to `Fabric8FlinkKubeClient`, but also about the concurrent issues. 

I just forgot to convert it the draft. That's why I didn't ping you for the review :).

**Comment by wangyang0918** [comment]
cc @tillrohrmann I have refined the second commit and I think this PR now is ready for review. Please have a look.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java [hunk 3]
```diff
@@ -70,17 +80,30 @@
...
 
+    /**
+     * Cache for Flink kubernetes clients. Key is master URL and namespace. Value is
+     * Fabric8FlinkKubeClient and reference count.
+     */
+    @GuardedBy("LOCK")
+    private static final Map<Tuple2<String, String>, Tuple2<Fabric8FlinkKubeClient, Integer>>
+            CACHE = new HashMap<>();
+    /** Lock used to protect modification of CACHE. */
+    private static final Object LOCK = new Object();
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:91]: Do we really need this complexity? What do we gain by it? We save instantiating 4 * x threads, right?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java [hunk 4]
```diff
@@ -70,17 +80,30 @@
...
     private final NamespacedKubernetesClient internalClient;
     private final String clusterId;
     private final String namespace;
     private final int maxRetryAttempts;
 
-    private final Executor kubeClientExecutorService;
+    private final ExecutorService kubeClientExecutorService;
+
+    /* F0 is master URL, and F1 is namespace. */
+    private final Tuple2<String, String> cacheKey;
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:101]: Let's not use `TupleX`. I think this class is very inexpressive.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java [hunk 5]
```diff
@@ -354,6 +394,29 @@ public KubernetesPod loadPodFromTemplateFile(File file) {
...
+                            KubernetesConfigOptions.KUBERNETES_CLIENT_IO_EXECUTOR_POOL_SIZE);
+            final ExecutorService executorService =
+                    Executors.newFixedThreadPool(
+                            poolSize, new ExecutorThreadFactory("flink-kubeclient-io@" + cacheKey));
+            final Fabric8FlinkKubeClient flinkKubeClient =
+                    new Fabric8FlinkKubeClient(flinkConfig, client, () -> executorService);
+            CACHE.put(cacheKey, Tuple2.of(flinkKubeClient, 1));
+            return flinkKubeClient;
+        }
+    }
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:418]: I would first try how it performs without the `CACHE`. I am sceptical whether this is really needed.

**Review by tillrohrmann — CHANGES_REQUESTED** [review]
Thanks for finalizing this PR @wangyang0918. The change of the `Fabric8FlinkKubeClient` looks good to me. What I am not really convinced is the `CACHE` mechanism. Why do you think that we need this mechanism?

**Comment by wangyang0918** [comment]
The reason why I introduce the `CACHE` for `Fabric8FlinkKubeClient` is about to reduce the number of threads. Also because `Fabric8FlinkKubeClient` is thread-safe, and then we have the opportunity to do this optimization.

I am not against with going without `CACHE` first and introduce it in the future if we really find that we need it. I will revert the second commit.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java [hunk 6]
```diff
@@ -342,7 +367,22 @@ public KubernetesWatch watchConfigMaps(
...
+        synchronized (LOCK) {
+            final Tuple2<Fabric8FlinkKubeClient, Integer> cachedClient = CACHE.get(cacheKey);
+            if (cachedClient != null) {
+                if (--cachedClient.f1 == 0) {
+                    cachedClient.f0.internalClose();
+                    CACHE.remove(cacheKey);
+                }
+            } else {
+                internalClose();
+            }
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:379]: If you really think this is not a correct behavior(instantiating `Fabric8FlinkKubeClient` does not use the `CACHE`), I could make the constructor `private`.

**Comment by wangyang0918** [comment]
cc @tillrohrmann The second commit has been reverted.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java [hunk 7]
```diff
@@ -75,12 +77,12 @@
...
     private final int maxRetryAttempts;
 
-    private final Executor kubeClientExecutorService;
+    private final ExecutorService kubeClientExecutorService;
 
     public Fabric8FlinkKubeClient(
             Configuration flinkConfig,
             NamespacedKubernetesClient client,
-            Supplier<Executor> asyncExecutorFactory) {
+            Supplier<ExecutorService> asyncExecutorFactory) {
```
Comment (tillrohrmann) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java:85]: Why don't we pass the `ExecutorService` directly in?

**Review by tillrohrmann — APPROVED** [review]
Thanks for updating this PR @wangyang0918. The changes look good to me. The last remaining comment I will address myself while merging.

**Pull request closed without merging**